hub/core/storage/s3.py

```diff
@@ -236,21 +258,52 @@ def _check_update_creds(self):
                 org_id, ds_name, mode
             )
             self.expiration = expiration
-            self.client = boto3.client(
-                "s3",
-                aws_access_key_id=creds["access_key"],
-                aws_secret_access_key=creds["secret_key"],
-                aws_session_token=creds["session_token"],
-                config=self.client_config,
-                endpoint_url=self.endpoint_url,
-                region_name=self.aws_region,
+            self._set_s3_client_and_resource(
+                creds.get("aws_access_key_id"),
+                creds.get("aws_secret_access_key"),
+                creds.get("aws_session_token"),
             )
-            self.resource = boto3.resource(
-                "s3",
-                aws_access_key_id=creds["access_key"],
-                aws_secret_access_key=creds["secret_key"],
-                aws_session_token=creds["session_token"],
-                config=self.client_config,
-                endpoint_url=self.endpoint_url,
-                region_name=self.aws_region,
+
+    def _locate_and_load_creds(self):
+        session = boto3._get_default_session()._session
+        component_locator = ComponentLocator()
+        component_locator.lazy_register_component(
+            "credential_provider", session._create_credential_resolver
+        )
+        credentials = component_locator.get_component(
+            "credential_provider"
+        ).load_credentials()
+        if credentials is not None:
+            self.aws_access_key_id = credentials.access_key
+            self.aws_secret_access_key = credentials.secret_key
+            self.aws_session_token = credentials.token
+            self.aws_region = session._resolve_region_name(
+                self.aws_region, self.client_config
             )
+
+    def _set_s3_client_and_resource(
+        self, aws_access_key_id=None, aws_secret_access_key=None, aws_session_token=None
+    ):
+        key = aws_access_key_id or self.aws_access_key_id
+        secret = aws_secret_access_key or self.aws_secret_access_key
+        token = aws_session_token or self.aws_session_token
+
+        self.client = boto3.client(
+            "s3",
+            aws_access_key_id=key,
+            aws_secret_access_key=secret,
+            aws_session_token=token,
+            config=self.client_config,
+            endpoint_url=self.endpoint_url,
+            region_name=self.aws_region,
+        )
+
+        self.resource = boto3.resource(
+            "s3",
+            aws_access_key_id=key,
+            aws_secret_access_key=secret,
+            aws_session_token=token,
+            config=self.client_config,
+            endpoint_url=self.endpoint_url,
+            region_name=self.aws_region,
+        )
```

Comment:
just curious, why do we have both resource and client? and why do they separately authenticate?

Reply:
resource is useful for doing bulk operations like 'clear'.